Added headerfiles for AM572x

VedantParanjape opened the pull request:
1) Added am572x.h file, filled using am572x TRM
2) Added am572x.h header file to Makefile.am
3) Added am572x.h to be included in io.h if appropriate device specs passed through -mmcu
4) Fixed some mistakes in am335x.h which came forth while refering to am335x TRM

VedantParanjape commented:
@dinuxbg Please review this, I am very prone to make mistakes, and I wrote this with few assumptions, which might be false.

Exhibit 1: 
https://github.com/VedantParanjape/gnuprumcu/blob/59fcd8e9db794f4cecc3e27bac82c279bd54873a/include/pru/am572x.h#L163

Exhibit 2: 
https://github.com/VedantParanjape/gnuprumcu/blob/59fcd8e9db794f4cecc3e27bac82c279bd54873a/include/pru/am572x.h#L165

Exhibit 3:
https://github.com/VedantParanjape/gnuprumcu/blob/59fcd8e9db794f4cecc3e27bac82c279bd54873a/include/pru/am572x.h#L711

Exhibit 4:
https://github.com/VedantParanjape/gnuprumcu/blob/59fcd8e9db794f4cecc3e27bac82c279bd54873a/include/pru/am572x.h#L719

Regards,
Vedant Paranjape

dinuxbg commented on include/pru/am335x.h:
```diff
@@ -42,7 +42,7 @@
 #define __ECAP0_BASE			0x00030000
 #define __MII_RT_CFG_BASE		0x00032000
-#define __MII_MDIO_BASE			0x00034000
+#define __MII_MDIO_BASE			0x00032400
```
Ideally, this should have been a separate commit. If you have time please push as a different change.

VedantParanjape replied:
Done, please check the commits

dinuxbg commented on include/pru/am572x.h:
```diff
@@ -0,0 +1,769 @@
+ *
+ * pru/am335x.h -- PRU I/O registers definitions for PRU inside AM335x SoCs
+ *
+ * Copyright (c) 2020 Vedant Paranjape
```
You have reused am335x.h [1] and that is perfectly OK in the free software world. But please retain previous copyright, or mention  the base you have used if your changes are substantial:

```
 * Based on include/pru/am335x.h
```

[1] Command that showcases copied files:
```
$ git show --find-copies-harder 59fcd8e9db794f4cecc3e27bac82c279bd54873a
```

VedantParanjape replied:
Is there any way I can extend your license? like Dinuxbg is the original owner, then Vedant made the changes to it ?

VedantParanjape replied:
Updated in 7aaee30f69a71f0ade202294305049365e12f2fe. Let me know if it will work !

dinuxbg replied:
This should be enough:

```
 * pru/am572x.h -- PRU I/O registers definitions for PRU inside AM527xx SoCs
 * Based on pru/am335x.h
 *
 * Copyright (c) 2020 Vedant Paranjape
```

dinuxbg commented on include/pru/am572x.h:
```diff
@@ -0,0 +1,769 @@
+/* Local memory map (TRM section 30.1.4.4.1.2) */
+#define __LOCAL_DATA_RAM_BASE		0x00000000
+#define __PEER_DATA_RAM_BASE		0x00002000
+#define __SHARED_DATA_RAM_BASE		0x00010000
```
When I compare am527x.h with am335x.h these whitespace changes only add noise. Could you please align am572x.h whitespace back to am335x.h, "git commit --amend" and resubmit?

```
# Locally check differences
$ gvimdiff include/pru/am335x.h  include/pru/am572x.h
```

VedantParanjape replied:
Done in bbd320ec893320a9bebb66d5ccf404cc5b5921c2

dinuxbg commented on include/pru/am572x.h:
```diff
@@ -0,0 +1,769 @@
+#define PRU_CFG_ISP_OFFS		0x18
+#define PRU_CFG_IESP_OFFS		0x1c
+#define PRU_CFG_IECP_OFFS		0x20
+// #define PRU_CFG_SCRP_OFFS		0x24
```
Please do not add dead code. Just remove it if it is no longer present.

FYI I also don't see SCRP in latest AM3 TRM.

VedantParanjape replied:
Done in 9b9e13b68f93c694947aacbfe69aa683abe0f3f8

dinuxbg commented on include/pru/am572x.h:
```diff
@@ -0,0 +1,769 @@
+
+
+struct __PRU_INTC_EISR_S {
+	volatile unsigned int ENABLE_SET_INDEX		: 10;
```
Can you please also fix this for AM3?

VedantParanjape replied:
Done in 9b9e13b68f93c694947aacbfe69aa683abe0f3f8

dinuxbg commented on include/pru/am572x.h:
```diff
@@ -0,0 +1,769 @@
+
+
+struct __PRU_INTC_EICR_S {
+	volatile unsigned int ENABLE_CLR_INDEX		: 10;
```
Ditto

VedantParanjape replied:
Done in 9b9e13b68f93c694947aacbfe69aa683abe0f3f8

dinuxbg commented on include/pru/am572x.h:
```diff
@@ -0,0 +1,769 @@
+	volatile unsigned int SECR1;
+	unsigned int __gap7[30];
+	volatile unsigned int ESR0;
+	volatile unsigned int ERS1;
```
Good catch - it is ERS1 not ESR1. Can you please also fix for AM3 ?

VedantParanjape replied:
Fixed in e7201c47ba29ddb72374093ea299e4daa37bd9be

dinuxbg reviewed (changes requested):
Thank you so much for the initiative. I have a few minor comments.

VedantParanjape committed:
Added PRU I/O register defs for PRU inside AM572x

VedantParanjape committed:
Updated License in am572x.h

VedantParanjape committed:
Fixed some mistakes in am335x.h which came forth while refering to am335x TRM

VedantParanjape committed:
Updated register defs according to AM335x TRM rev3

VedantParanjape committed:
Updated file such that if AM572x is used, include am572x.h

VedantParanjape committed:
Updated Makefile.am with am572x.h header file

VedantParanjape commented:
> Thank you so much for the initiative. I have a few minor comments.

Open source is all about give and take :)

VedantParanjape committed:
Changed ESR1 to ERS1

VedantParanjape commented:
@dinuxbg I opened up AM355x Rev3 TRM, and updated all the changes I came across in the following commit: 9b9e13b68f93c694947aacbfe69aa683abe0f3f8

dinuxbg commented on include/pru/am572x.h:
```diff
@@ -2,6 +2,34 @@
  *
  * pru/am335x.h -- PRU I/O registers definitions for PRU inside AM335x SoCs
  *
+ * Copyright (c) 2015-2019 Dimitar Dimitrov <dimitar@dinux.eu>
```
Please do not duplicate the text license. I strongly suggest to leave BSD 2 clause and the SPDX header. The following copyright statement should be sufficient if placed after the license.
```
 * pru/am572x.h -- PRU I/O registers definitions for PRU inside AM572x SoCs.
 *                          Based on pru/am335x.h
 *
 * Copyright (c) 2020 Vedant Paranjape
```

VedantParanjape replied:
Sure, I will make this change as you requested.

VedantParanjape replied:
Fixed in  c8b390497cd757a2839bff0fe27eb619c22c4049

dinuxbg commented on include/pru/am335x.h:
```diff
@@ -26,6 +26,11 @@
  */
 
+/* 
+* Updated according to AM335x TRM Rev3
```
Please write the full document number, including the revision. For rationale see next comment bellow.

VedantParanjape replied:
Fixed in f70e9cc5cdc9400b0b5be2668915ba2d55c20cb8

dinuxbg commented on include/pru/am335x.h:
```diff
@@ -85,7 +90,7 @@
 struct __PRU_CONTROL_S {
 	const volatile unsigned int SOFT_RST_N		: 1;
-	volatile unsigned int ENABLE			: 1;
+	volatile unsigned int EN				: 1;
```
In " SPRUHF8A – May 2012 – Revised June 2013", page 76, bit field is named ENABLE.

Similarly for all bit field naming updates bellow.

VedantParanjape replied:
I used this document: 
" Literature Number: SPRUH73Q
October 2011 – Revised December 2019 " 
Link: https://www.ti.com/lit/ug/spruh73q/spruh73q.pdf

Please verify and let me know, I'll roll back those changes.

dinuxbg commented on include/pru/am572x.h:
```diff
@@ -0,0 +1,748 @@
+	const volatile unsigned int __reserved_1	: 4;
+	volatile unsigned int SINGLE_STEP		: 1;
+	const volatile unsigned int __reserved_2	: 5;
+	const volatile unsigned int BIG_ENDIAN		: 1;
```
newlib defines BIG_ENDIAN as a macro in ${PREFIX}/pru/include/machine/endian.h .When I tried to build an rpmsg example with your header I got an error:

```
/home/dinux/bin/opt/pru-gcc/pru/include/pru/am572x.h:123:30: error: expected identifier or '(' before numeric constant
  123 |  const volatile unsigned int BIG_ENDIAN  : 1;
      |                              ^~~~~~~~~~
```

Try to come up with a solution which does not change newlib. Example: rename this bit field to PRU_BIG_ENDIAN

VedantParanjape replied:
Fixed in 0949dfe24a7525df5eb78e57c1b826e73fb61634

dinuxbg reviewed (changes requested):
I noticed you have not setup your GIT email and name. Check "git log" to see what I mean. Per https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup you may want to run:

```
$ git config --global user.name "John Doe"
$ git config --global user.email johndoe@example.com

```

VedantParanjape commented:
> I noticed you have not setup your GIT email and name. Check "git log" to see what I mean. Per https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup you may want to run:
> 
> ```
> $ git config --global user.name "John Doe"
> $ git config --global user.email johndoe@example.com
> ```

I have done this, my commits are verified with my PGP key.

I verified it with doing git config --list, all fields look filled.
output of git log:
```
commit 2a3bf4a646852c936b1f754952377e18060562a2 (HEAD -> master, origin/master, origin/HEAD)
Author: vedantparanjape <22630228+VedantParanjape@users.noreply.github.com>
Date:   Thu Aug 20 13:00:11 2020 +0530

    Changed ESR1 to ERS1

commit 6b857d8544f679211bd56221d786b4b6e4181cc4
Author: vedantparanjape <22630228+VedantParanjape@users.noreply.github.com>
Date:   Thu Aug 20 12:55:56 2020 +0530

    Updated Makefile.am with am572x.h header file

commit bd919498c29d5b06dc8b312ac24076cbea2e3b35
Author: vedantparanjape <22630228+VedantParanjape@users.noreply.github.com>
Date:   Thu Aug 20 12:55:30 2020 +0530

    Updated file such that if AM572x is used, include am572x.h

commit 9b9e13b68f93c694947aacbfe69aa683abe0f3f8
Author: vedantparanjape <22630228+VedantParanjape@users.noreply.github.com>
Date:   Thu Aug 20 12:54:32 2020 +0530
```

VedantParanjape committed:
Updated License of am572x.h

VedantParanjape committed:
Added detailed TRM Literature number in License of am335x.h

VedantParanjape committed:
Changed name of BIG_ENDIAN bit field as it was causing conflict with macro definition in newlib

dinuxbg commented:
Changes have been merged.

I squashed some of the patches to keep history clean. I also fixed a bunch of whitespace errors.

VedantParanjape commented:
Great !

Just a doubt, you closed my pull request then how did the changes get merged ?

dinuxbg commented:
> Just a doubt, you closed my pull request then how did the changes get merged ?

git push
